lucille-core/src/main/java/com/kmwllc/lucille/stage/FetchUri.java

```diff
@@ -0,0 +1,112 @@
+package com.kmwllc.lucille.stage;
+
+import com.kmwllc.lucille.core.ConfigUtils;
+import com.kmwllc.lucille.core.Document;
+import com.kmwllc.lucille.core.Stage;
+import com.kmwllc.lucille.core.StageException;
+import com.typesafe.config.Config;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.Iterator;
+import nl.altindag.ssl.util.UriUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.input.BoundedInputStream;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.ClientProtocolException;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+
+public class FetchUri extends Stage {
+
+  private final String source;
+  private final String dest;
+  private final String statusSuffix;
+  private final String sizeSuffix;
+
+  private final String errorSuffix;
+  private final int maxDownloadSize;
+  private CloseableHttpClient client;
+
+  public FetchUri(Config config) {
+    super(config, new StageSpec().withRequiredProperties("source", "dest")
+        .withOptionalProperties("size_suffix", "status_suffix", "max_size", "error_suffix"));
+    this.source = config.getString("source");
+    this.dest = config.getString("dest");
+    this.statusSuffix = ConfigUtils.getOrDefault(config, "status_suffix", "status_code");
+    this.sizeSuffix = ConfigUtils.getOrDefault(config, "size_suffix", "size");
+    this.errorSuffix = ConfigUtils.getOrDefault(config, "error_suffix", "error");
+    this.maxDownloadSize = ConfigUtils.getOrDefault(config, "max_size", Integer.MAX_VALUE);
+  }
+
+  void setClient(CloseableHttpClient client) {
+    this.client = client;
+  }
+
+  @Override
+  public void start() throws StageException {
+    client = HttpClients.createDefault();
+  }
+
+  @Override
+  public Iterator<Document> processDocument(Document doc) throws StageException {
+    if (!doc.has(source) || doc.getString(source).isEmpty()) {
+      return null;
+    }
+    String uri = doc.getString(source);
+
+    // Following checks for first valid URL and if not, then valid URI, and then if not places error message in error field
```

Comment:
A URI can be a URL or URN. I don't think we need to try to support URNs here, so it should be sufficient to simply require a valid URL.

Reply:
Changed the validation to now only check for valid URL, updated comment as well